acstools/acszpt.py

```diff
@@ -0,0 +1,329 @@
+        else:
+            self.filt = filt.upper()
+
+            self._url = 'https://acszeropoints.stsci.edu/results_single/' \
+                        '?date1={0}&detector={1}&{1}_filter={2}'. \
+                format(self.date, self.detector, self.filt)
+
+        self._zpts = OrderedDict()
+        self._response = None
+        self._failed = False
+
+    def _check_inputs(self):
+        """ Check the input to ensure they are valid
+
+        Returns
+        -------
+
+        """
+        valid_date = True
+        valid_detector = True
+        valid_filter = True
+        # Determine the submitted detector is valid
+        if self.detector not in ['HRC','SBC','WFC']:
+            print('Whoops! {} is not a valid detector option.'.
+                  format(self.detector))
+            print('Please choose one of the following:')
+            print('HRC')
+            print('SBC')
+            print('WFC')
+            print('-'*79)
+            valid_detector = False
+
+        # Determine if the submitted filter is valid
+        if self.filt is None:
+            pass
+        elif self.filt not in self.valid_filters[self.detector]:
+            valid_filter = False
+            print('Whoops! {} is not a valid filter for {}'.
+                  format(self.filt, self.detector))
+            print('Please choose one of the following:')
+            for filt in self.valid_filters[self.detector]:
+                print(filt)
+            print('-' * 79)
+
+        # Determine if the submitted date is valid
+        date_check = self._check_date()
+        if isinstance(date_check, ValueError):
+            valid_date = False
+            print('Whoops! {}'.format(date_check))
+
+        # Only continute if all the inputs are valid options
+        if all([valid_date, valid_filter, valid_detector]):
+            return True
+        else:
+            return False
+
+    def _check_date(self, fmt='%Y-%m-%d'):
+        """Convenience method for determing if the input date is valid
+        """
+        min_date = dt.datetime(2002, 3, 7)
+        try:
+            dt_obj = dt.datetime.strptime(self.date, fmt)
+        except ValueError:
+            return ValueError('Date format does not match YYYY-MM-DD')
+        else:
+            if min_date < dt_obj:
+                result = None
+            else:
+                result = ValueError('The observation cannot occur before '
+                                 'ACS was installed ({})'.
+                                 format(min_date.strftime(fmt)))
+            return result
+
+    def _submit_request(self):
+        """ Submit a request to the ACS Zeropoint Calculator
+
+        If an exception is raised during the request an error message is
+        printed. Otherwise, the response is saved in the corresponding
+        attribute.
+        """
+        try:
+            self._response = requests.get(self._url)
+        except requests.exceptions.RequestException as e:
```

Comment:
Are you sure it raises an exception at all? http://docs.python-requests.org/en/master/

I thought its behavior is to store failure inside `self._response.status_code`?

Reply:
I _think_ the only case that it will raise an exception for is when the server is down and it exceeds that max retries. 
```console
In [4]: q = acszpt.Query(detector='WFC', date='2004-02-01')                               

In [5]: q.fetch()                                                                         
Out[5]: 
<Table length=13>
FILTER PHOTPLAM        PHOTFLAM         STmag  VEGAmag  ABmag 
       Angstrom erg / (Angstrom cm2 s) mag(ST)   mag   mag(AB)
 str6  float64         float64         float64 float64 float64
------ -------- ---------------------- ------- ------- -------
 F435W   4328.8               3.07e-19  25.182   25.79  25.692
 F475W   4745.2              1.786e-19   25.77  26.173  26.081
 F502N   5023.0              5.148e-18  22.121  22.388  22.308
 F550M   5581.3              3.921e-19  24.917  24.844  24.875
 F555W   5360.1              1.927e-19  25.688  25.734  25.734
 F606W   5919.3              7.694e-20  26.685  26.423  26.515
 F625W   6311.1              1.172e-19  26.227  25.749  25.919
 F658N   6584.0              1.947e-18  23.176  22.394  22.776
 F660N   6599.4              5.096e-18  22.132  21.441  21.726
 F775W   7693.5              9.849e-20  26.417  25.284  25.678
 F814W   8047.1              6.957e-20  26.794   25.53  25.958
F850LP   9036.5              1.491e-19  25.966  24.352  24.878
 F892N   8914.8              1.477e-18  23.477  21.923  22.418

In [6]: q._url                                                                            
Out[6]: 'http://127.0.0.1:5000/results_all/?date=2004-02-01&detector=WFC'
```
Now I shut down the development server and attempt the same query, 
```console
In [7]: q.fetch()                                                                         
HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url: /results_all/?date=2004-02-01&detector=WFC (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x1c216757b8>: Failed to establish a new connection: [Errno 61] Connection refused',))
-------------------------------------------------------------------------------
The query failed! Check your input args and if the error persists submit a ticket to the ACS Help Desk at hsthelp.stsci.edu with the error message displayed above.
```
I'm glad you pointed this out because for some reason I thought the `requests` module was a native python module. I switched everything to use `urllib.request` and its corresponding errors module `urllib.error` (no need for unnecessary dependencies). The result is the same in that if the server is down it will raise an exception stating max retries, so the code is unchanged.